text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,604 @@
+`const`. Since `Add` only has `add` as a method, in this case one only needs to ensure that the `add` method is
+`const`. Instead of adding a `const` modifier to all methods of a trait impl, the modifier is added to the entire
+`impl` block:
+
+```rust
+struct MyInt(i8);
+impl const Add for MyInt {
+    fn add(self, other: Self) -> Self {
+        MyInt(self.0 + other.0)
+    }
+}
+```
+
+You cannot implement both `const Add` and `Add` for any type, since the `const Add`
+impl is used as a regular impl outside of const contexts.
+
+The const requirement is inferred on all bounds of the impl and its methods,
+so in the following `H` is required to have a const impl of `Hasher`, so that
+methods on `state` are callable.
+
+```rust
+impl const Hash for MyInt {
+    fn hash<H>(
+        &self,
+        state: &mut H,
+    )
+        where H: Hasher
+    {
+        state.write(&[self.0 as u8]);
+    }
+}
+```
+
+The same goes for associated types' bounds: all the bounds require `impl const`s for the type used
+for the associated type:
+
+```rust
+trait Foo {
+    type Bar: Add;
+}
+impl const Foo for A {
+    type Bar = B; // B must have an `impl const Add for B`
+}
+```
+
+If an associated type has no bounds in the trait, there are no restrictions to what types may be used
+for it.
+
+These rules for associated types exist to make this RFC forward compatible with adding const default bodies
+for trait methods. These are further discussed in the "future work" section.
+
+## Generic bounds
+
+The above section skimmed over a few topics for brevity. First of all, `impl const` items can also
+have generic parameters and thus bounds on these parameters, and these bounds follow the same rules
+as bounds on generic parameters on `const` functions: all bounds can only be substituted with types
+that have `impl const` items for all the bounds. Thus the `T` in the following `impl` requires that
+when `MyType<T>` is used in a const context, `T` needs to have an `impl const Add for Foo`.
+
+```rust
+impl<T: Add> const Add for MyType<T> {
+    /* some code here */
+}
+const FOO: MyType<u32> = ...;
+const BAR: MyType<u32> = FOO + FOO; // only legal because `u32: const Add`
+```
+
+Furthermore, if `MyType` is used outside a const context, there are no constness requirements on the
+bounds for types substituted for `T`.
+
+## Drop
+
+A notable use case of `impl const` is defining `Drop` impls.
+Since const evaluation has no side effects, there is no simple example that
+showcases `const Drop` in any useful way. Instead we create a `Drop` impl that
+has user visible side effects:
+
+```rust
+let x = Cell::new(42);
+SomeDropType(&x);
+// x is now 41
+
+struct SomeDropType<'a>(&'a Cell<u32>);
+impl const Drop for SomeDropType {
+    fn drop(&mut self) {
+        self.0.set(self.0.get() - 1);
+    }
+}
+```
+
+You are now allowed to actually let a value of `SomeDropType` get dropped within a constant
```

Comment:
So... ignoring this RFC: even if you have a `T: Drop` bound on a function, you can still pass `u32` as that `T` right now

With this RFC, the "logical conclusion" from allowing `T: Add` to run `T::add` at const eval time is that we need to specify `T: Drop` to have the compiler run `T::drop` at the end of the appropriate scope.

Since adding a `Drop` bound to a generic parameter does neither change anything in terms of what you can do inside the function with the generic parameter nor change what types you can pass, this is a valid option.

> that seems like it requires a `Drop` impl, which is wrong

Well, it only seems like it, it does not actually require a *manual* `Drop` impl. Every type implements `Drop`.

Reply:
It *is* misleading, but I don't see any other way that is consistent with everything else in this RFC. Suggestions are very welcome.

Reply:
I personally mildly prefer `const drop`, to indicate that this is "special" and not about the literal `Drop` trait, but I don't feel super strongly about it

Reply:
We could also do something like have a marker trait `ConstDrop` which is automatically implemented for all types that have an `impl const Drop` or are just aggregates of `ConstDrop` or `Copy` types.

To start out we could make it illegal to implement `ConstDrop` manually for your type, but in the future we could consider something like `unsafe impl ConstDrop for MyType {}` which would allow you to have a non-const `Drop` impl for runtime and just do nothing at compile-time.

Reply:
@oli-obk: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4e054f337767842c3f4d57251790fbc0

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: the trait bound `i32: std::ops::Drop` is not satisfied
 --> src/main.rs:4:5
  |
4 |     foo(1i32);
  |     ^^^ the trait `std::ops::Drop` is not implemented for `i32`
  |
note: required by `main::foo`
 --> src/main.rs:3:5
  |
3 |     fn foo<T: Drop>(x: T) {}
  |     ^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Oh wow. I am very confused now. I don't know why I ever thought that worked.

Ok... in that case we definitely do need something like `ConstDrop`

Reply:
Another option would be to disallow instantiating types that have a non-`const Drop`, thus the only values that could be passed into the function are guaranteed to be `const Drop`. Especially since we don't have `unsafe` in `const` contexts.

Reply:
That would be problematic for e.g. `new` methods on generic types:

```rust
struct Foo<T>(T);
impl Foo<T> {
    const fn new(t: T) -> Self { Self(t) }
}
```

would be forbidden by your suggested rule I think.

Reply:
Only if `Foo` has a non-`const Drop`. This is because we can assume that `T` jas a `const Drop` because we have an instance of it, and the only qay we qould have an instance of it in a const context is if it has a `const Drop`.

Reply:
That limitation would not only be backwards incompatible, it would also be very restrictive. I think we must find a way to keep the example working.

Reply:
There is no back-compat hazard because destructors cannot be evaluated at compile time right now (i.e. in a `const` context)

[proof](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1143e7c3b4546d1ff208e7c2dcf98cdb)

So allowing const-Drop would actually allow *more* types to be passed in.

Reply:
If we can prove that the function's body does not call `Drop`, we do allow this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a740945359490c4c0efda5fae89bf639

Reply:
Oh, you edited your comment. So going back to my example above

```rust
struct Bar<T>(T);
impl Bar<T> {
    const fn new(t: T) -> Self { Self(t) }
}
```

This is already perfectly allowed on stable Rust. Merging this with your example we'd get

```rust
struct Foo;
struct Bar(Foo);

impl Drop for Foo {
    fn drop(&mut self) {
        println!("Hello");
    }
}

impl Bar {
    const fn new(foo: Foo) -> Self { Self(foo) }
}
```

which is also already legal on stable. This is not a problem, since [adding a `Drop` impl to a type is already a breaking change](https://github.com/rust-lang/rust/issues/58379).

Reply:
Oh, I wasn't aware of that, thanks!